File: app/src/main/java/com/xadaptersimple/MultipleItemActivity.java
package com.xadaptersimple;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import com.xadapter.adapter.multi.MultiAdapter;
import com.xadapter.adapter.multi.MultiCallBack;
import com.xadapter.adapter.multi.SimpleMultiItem;
import com.xadapter.adapter.multi.XMultiAdapterListener;
import com.xadapter.holder.XViewHolder;
import com.xadapter.listener.OnItemClickListener;
import com.xadapter.listener.OnItemLongClickListener;

import java.util.ArrayList;
import java.util.List;

/**
 * by y on 2017/1/12.
 */

public class MultipleItemActivity extends AppCompatActivity
        implements
        OnItemClickListener<SimpleMultiItem>,
        OnItemLongClickListener<SimpleMultiItem>, XMultiAdapterListener<SimpleMultiItem> {
    private static final int TYPE_LINE = 1;

    public static List<SimpleMultiItem> initSettingData() {
        List<SimpleMultiItem> list = new ArrayList<>();
        list.add(new SimpleMultiItem(TYPE_LINE));
        list.add(new SimpleMultiItem(SimpleMultiItem.TYPE_ITEM, 0, "头像", R.mipmap.ic_launcher));
        list.add(new SimpleMultiItem(TYPE_LINE));
        list.add(new SimpleMultiItem(SimpleMultiItem.TYPE_ITEM, 1, "收藏", R.mipmap.ic_launcher));
        list.add(new SimpleMultiItem(SimpleMultiItem.TYPE_ITEM, 2, "相册", R.mipmap.ic_launcher));
        list.add(new SimpleMultiItem(TYPE_LINE));
        list.add(new SimpleMultiItem(SimpleMultiItem.TYPE_ITEM, 3, "钱包", R.mipmap.ic_launcher));
        list.add(new SimpleMultiItem(SimpleMultiItem.TYPE_ITEM, 4, "卡包", R.mipmap.ic_launcher));
        list.add(new SimpleMultiItem(TYPE_LINE));
        list.add(new SimpleMultiItem(SimpleMultiItem.TYPE_ITEM, 5, "表情", R.mipmap.ic_launcher));
        list.add(new SimpleMultiItem(TYPE_LINE));
        list.add(new SimpleMultiItem(SimpleMultiItem.TYPE_ITEM, 6, "设置", R.mipmap.ic_launcher));
        return list;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.multiple_layout);
        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recyclerView);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.addItemDecoration(new DividerItemDecoration(this, LinearLayoutManager.VERTICAL));

        recyclerView.setAdapter(
                new MultiAdapter<>(initSettingData())
                        .setXMultiAdapterListener(this)
                        .setOnItemClickListener(this)
                        .setOnLongClickListener(this));
    }

    @Override
    public void onItemClick(View view, int position, SimpleMultiItem info) {
        Toast.makeText(view.getContext(), "当前 position:  " + position + "  " + info.message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public boolean onLongClick(View view, int position, SimpleMultiItem info) {
        Toast.makeText(view.getContext(), "当前内容  = " + info.message, Toast.LENGTH_SHORT).show();
        return true;
    }

    @Override
    public int multiLayoutId(int viewItemType) {
        switch (viewItemType) {
            case TYPE_LINE:
                return R.layout.item_line;
            default:
                return R.layout.item_multi;
        }
    }

    @Override
    public int getGridLayoutManagerSpanSize(int itemViewType, GridLayoutManager gridManager, int position) {
        if (!(itemViewType == MultiCallBack.TYPE_ITEM)) {
            return gridManager.getSpanCount();
        } else {
            return 1;
        }
    }

    @Override
    public boolean getStaggeredGridLayoutManagerFullSpan(int itemViewType) {
        return itemViewType != MultiCallBack.TYPE_ITEM;
    }

    @Override
    public void onXMultiBind(XViewHolder holder, SimpleMultiItem simpleMultiItem, int itemViewType, int position) {
        switch (itemViewType) {
            case SimpleMultiItem.TYPE_ITEM:
                holder.setTextView(R.id.tv_message, simpleMultiItem.message);
                ImageView imageView = holder.getImageView(R.id.iv_icon);
                imageView.setImageResource(simpleMultiItem.icon);
                break;
        }
    }
}

File: app/src/main/java/com/xadaptersimple/MultipleItemActivity.kt
package com.xadaptersimple

import android.os.Bundle
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.xadapter.OnItemClickListener
import com.xadapter.OnItemLongClickListener
import com.xadapter.OnXMultiAdapterListener
import com.xadapter.adapter.XMultiAdapter
import com.xadapter.holder.XViewHolder
import com.xadapter.simple.SimpleXMultiItem
import kotlinx.android.synthetic.main.recyclerview_layout.*

/**
 * by y on 2017/1/12.
 */

class MultipleItemActivity : AppCompatActivity(), OnItemClickListener<SimpleXMultiItem>,
        OnItemLongClickListener<SimpleXMultiItem>,
        OnXMultiAdapterListener<SimpleXMultiItem> {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.multiple_layout)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.addItemDecoration(DividerItemDecoration(this, LinearLayoutManager.VERTICAL))
        recyclerView.adapter = XMultiAdapter(initSettingData())
                .apply {
                    onXMultiAdapterListener = this@MultipleItemActivity
                    onItemClickListener = this@MultipleItemActivity
                    onLongClickListener = this@MultipleItemActivity
                }
    }

    override fun onItemClick(view: View, position: Int, entity: SimpleXMultiItem) {
        Toast.makeText(view.context, "当前 position:  " + entity.itemMultiPosition + "  " + entity.message, Toast.LENGTH_SHORT).show()
    }

    override fun onLongClick(view: View, position: Int, entity: SimpleXMultiItem): Boolean {
        Toast.makeText(view.context, "当前内容  = " + entity.message, Toast.LENGTH_SHORT).show()
        return true
    }

    override fun multiLayoutId(viewItemType: Int): Int {
        return when (viewItemType) {
            TYPE_LINE -> R.layout.item_line
            else -> R.layout.item_multi
        }
    }

    override fun getGridLayoutManagerSpanSize(itemViewType: Int, gridManager: GridLayoutManager, position: Int): Int {
        return if (itemViewType != TYPE_ITEM) {
            gridManager.spanCount
        } else {
            1
        }
    }

    override fun getStaggeredGridLayoutManagerFullSpan(itemViewType: Int): Boolean {
        return itemViewType != TYPE_ITEM
    }

    override fun onXMultiBind(holder: XViewHolder, entity: SimpleXMultiItem, itemViewType: Int, position: Int) {
        when (itemViewType) {
            TYPE_ITEM -> {
                holder.setTextView(R.id.tv_message, entity.message)
                val imageView = holder.getImageView(R.id.iv_icon)
                imageView.setImageResource(entity.icon)
            }
        }
    }

    companion object {
        private const val TYPE_LINE = 1
        private const val TYPE_ITEM = -11
        fun initSettingData(): MutableList<SimpleXMultiItem> {
            return ArrayList<SimpleXMultiItem>().apply {
                add(SimpleXMultiItem(itemMultiType = TYPE_LINE))
                add(SimpleXMultiItem(itemMultiType = TYPE_ITEM, itemMultiPosition = 0, message = "头像", icon = R.mipmap.ic_launcher))
                add(SimpleXMultiItem(itemMultiType = TYPE_LINE))
                add(SimpleXMultiItem(itemMultiType = TYPE_ITEM, itemMultiPosition = 1, message = "收藏", icon = R.mipmap.ic_launcher))
                add(SimpleXMultiItem(itemMultiType = TYPE_ITEM, itemMultiPosition = 2, message = "相册", icon = R.mipmap.ic_launcher))
                add(SimpleXMultiItem(itemMultiType = TYPE_LINE))
                add(SimpleXMultiItem(itemMultiType = TYPE_ITEM, itemMultiPosition = 3, message = "钱包", icon = R.mipmap.ic_launcher))
                add(SimpleXMultiItem(itemMultiType = TYPE_ITEM, itemMultiPosition = 4, message = "卡包", icon = R.mipmap.ic_launcher))
                add(SimpleXMultiItem(itemMultiType = TYPE_LINE))
                add(SimpleXMultiItem(itemMultiType = TYPE_ITEM, itemMultiPosition = 5, message = "表情", icon = R.mipmap.ic_launcher))
                add(SimpleXMultiItem(itemMultiType = TYPE_LINE))
                add(SimpleXMultiItem(itemMultiType = TYPE_ITEM, itemMultiPosition = 6, message = "设置", icon = R.mipmap.ic_launcher))
            }
        }
    }
}
